### PR TITLE
Proxy db

### PR DESCRIPTION
**Pull request by bitdivine**
# Motivation
The plan for migrating data involves having one or more account databases and a virtual database that sits on top of these other databases and deals with the migration and database calls during the migration.

# Changes
- Define a proxy DB.

# Tests
- The standard db tests are applied to the proxy DB.

# Todos

- [x] Add entry to changelog (if necessary).

### CODE REVIEW

File: rs/backend/src/accounts_store/schema/map.rs [hunk 1]
```diff
@@ -30,6 +31,26 @@ impl AccountsDbTrait for AccountsDbAsMap {
     }
 }
 
+impl AccountsDbAsMap {
+    /// Creates a db from a map of accounts.
```
Comment (bitdivine) [inline rs/backend/src/accounts_store/schema/map.rs:35]: In the `post_upgrade` hook, a large blob of candid is parsed and the result includes a hashmap of all accounts.  `from_map()` lets us create a database from that data.

File: rs/backend/src/accounts_store/schema/map.rs [hunk 2]
```diff
@@ -30,6 +31,26 @@ impl AccountsDbTrait for AccountsDbAsMap {
...
+    #[allow(dead_code)] // TODO: Remove allow when this is used in production.
+    pub fn from_map(map: BTreeMap<Vec<u8>, Account>) -> Self {
+        Self { accounts: map }
+    }
+    /// Provides the DB contents as a map.
```
Comment (bitdivine) [inline rs/backend/src/accounts_store/schema/map.rs:40]: This is used in the pre-upgrade hook when serializing data to candid.

File: rs/backend/src/accounts_store/schema/proxy.rs [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+#[derive(Default, Debug)]
+pub struct AccountsDbAsProxy {
+    map: AccountsDbAsMap,
+}
+
+impl AccountsDbAsProxy {
+    /// Creates a db from a hash map of accounts.
+    #[allow(dead_code)] // TODO: Remove allow when this is used in production.
+    pub fn from_map(map: BTreeMap<Vec<u8>, Account>) -> Self {
+        Self {
```
Comment (bitdivine) [inline rs/backend/src/accounts_store/schema/proxy.rs:28]: Note: Currently we have only one underlying implementation, so we can proxy calls straight through to the underlying `AccountsDbAsMap`.

File: rs/backend/src/accounts_store/schema/map.rs (hunk 2 above)
Comment (frederikrothenberger) [reply rs/backend/src/accounts_store/schema/map.rs:40]: Shouldn't this be used somewhere as part of a test allowing you to replace `#[allow(dead_code)]` with `#[cfg(test)]`?

File: rs/backend/src/accounts_store/schema/proxy.rs [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+///
+/// Notes:
+/// - The proxy manages migrations from one implementation to another, if applicable.
+/// - The proxy code itself will specify which databases are currently in
+///   use and how to migrate from one database to another.
+/// - It is the responsibility of the post-install hook to look at any
+///   version information and set up the db accordingly.
+///
+/// # Current data storage
+/// - Accounts are
```
Comment (frederikrothenberger) [inline rs/backend/src/accounts_store/schema/proxy.rs:18]: There seems to be something missing here?

**Review by frederikrothenberger — APPROVED** [review]
Not a huge fan of the `#[allow(dead_code)]` attributes. I think it would be better to use these methods in tests too.

**Comment by bitdivine** [comment]
@frederikrothenberger Thank you for the review.  I put the `from_map()` and `as_map()` methods in a trait, which makes the complaints about dead code go away.  I also tested that trait on both the proxy  and map databases.  You have approved, but in view of the changes you may want to have another look.

**Review by frederikrothenberger — APPROVED** [review]
Thanks, LGTM!

**Pull request merged**